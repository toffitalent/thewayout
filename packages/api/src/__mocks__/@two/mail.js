const send = jest.fn().mockResolvedValue(undefined);

module.exports = {
  send,
};
