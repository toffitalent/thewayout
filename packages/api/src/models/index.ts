export * from './App';
export * from './AppClient';
export * from './AppToken';
export * from './AppUser';
export * from './BaseModel';
export * from './Client';
export * from './Employer';
export * from './Job';
export * from './JobApplication';
export * from './Rsp';
export * from './RspAccount';
export * from './RspInvitation';
export * from './serializers';
export * from './User';
export {
  CheckViolationError,
  ConstraintViolationError,
  DataError,
  DBError,
  ForeignKeyViolationError,
  NotFoundError,
  NotNullViolationError,
  raw,
  ref,
  UniqueViolationError,
  ValidationError,
} from 'objection';
