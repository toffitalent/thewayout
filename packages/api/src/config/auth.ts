import { env } from '@two/config';
import { UserType } from '@two/shared';

export const SCOPES = ['user'];

// Roles CANNOT be requested! Roles are automatically expanded to include all scopes specified.
// IMPORTANT: Ensure role names are UNIQUE and DO NOT CONFLICT with scope names!!!
export const ROLES = {
  admin: [UserType.Client, UserType.Employer, UserType.Rsp, ...SCOPES],
  self: [...SCOPES],

  [UserType.Client]: [],
  [UserType.Employer]: [],
  [UserType.Rsp]: [],
};

export const auth = () => {
  const auth = {
    algorithm: 'HS256',
    audience: 'two-api',
    issuer: 'two',
    secret: env({
      development:
        'MGICAQACEQC0GLtyi4/+XKuxzUliJ0B1AgMBAAECECyOaeG+8ZMGQ6jVVls5GIECCQDrhO4++DizMQIJAMPB/yr+OqiFAgg4tVU4TEQ6wQIIdGJ+ht0nmtECCQDHw8QHR1TrFA==',
      test: 'MGICAQACEQC0GLtyi4/+XKuxzUliJ0B1AgMBAAECECyOaeG+8ZMGQ6jVVls5GIECCQDrhO4++DizMQIJAMPB/yr+OqiFAgg4tVU4TEQ6wQIIdGJ+ht0nmtECCQDHw8QHR1TrFA==',
      env: 'API_AUTH_SECRET',
    }),
    ttl: 900,

    scopes: SCOPES,
    roles: ROLES,
  };

  const authMiddleware = {
    algorithm: auth.algorithm,
    audience: auth.audience,
    issuer: auth.issuer,
    roles: auth.roles,
    secret: auth.secret,
  };

  return { auth, server: { auth: authMiddleware } };
};
