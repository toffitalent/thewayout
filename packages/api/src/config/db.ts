import { env, register } from '@two/config';

export const db = register('db', () => ({
  hostname: env('API_DB_HOSTNAME', 'localhost'),
  username: env('API_DB_USERNAME', 'two'),
  password: env('API_DB_PASSWORD', ''),
  database: env('API_DB_DATABASE', 'two'),
  debug: env('API_DB_DEBUG', false),
  port: env('API_DB_PORT', 5432),
}));
