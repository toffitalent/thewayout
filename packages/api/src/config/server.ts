import { env, register } from '@two/config';

export const server = register('server', () => ({
  port: env({
    default: 3000,
    env: 'PORT',
    prefix: false,
  }),
  proxy: true,

  log: env({
    default: {
      level: 'info',
    },
    development: {
      level: 'debug',
      colors: true,
    },
    test: {
      level: 'none',
    },
  }),

  store: {
    url: env({
      default: 'redis://localhost:6379',
      env: 'API_REDIS_URL',
    }),
  },

  // Middleware
  auth: {
    //
    // NOTE: These settings are overridden in the "auth" config file
    //
    // enabled: true,
    // algorithm: 'HS256',
    // roles: [],
    // secret: '',
  },
  cache: {
    // enabled: true,
    // maxAge: 0,
  },
  cors: {
    // enabled: true,
    // origin: '',
    // allowHeaders: [],
    // allowMethods: [],
    // exposeHeaders: [],
    // maxAge: number,
  },
  csp: {
    enabled: false,
    // directives: {},
  },
  cto: {
    // enabled: true,
    // directive: '',
  },
  gzip: {
    // enabled: true
  },
  hsts: env({
    default: {
      // enabled: true,
      // maxAge: 0,
      // includeSubdomains: true,
      // preload: false,
    },
    development: {
      enabled: false,
    },
    staging: {
      maxAge: 3600,
    },
    production: {
      maxAge: 31536000,
    },
  }),
  json: {
    // enabled: true,
  },
  logger: env({
    default: {
      // enabled: true,
      // format: ':method :url :status :response-time ms - :length',
    },
    test: {
      enabled: false,
    },
    staging: {
      format:
        ':id :ip :app :user [:date] ":method :url" :status :length :response-time ":referrer" ":user-agent"',
    },
    production: {
      format:
        ':id :ip :app :user [:date] ":method :url" :status :length :response-time ":referrer" ":user-agent"',
    },
  }),
  methodOverride: {
    // enabled: true,
  },
  parser: {
    // enabled: true,
    // encoding = 'utf-8',
    // formLimit = '56kb',
    // json = true,
    // jsonLimit = '1mb',
    // multipart = false,
    // parsedMethods = ['POST', 'PUT', 'PATCH'],
    // text = false,
    // textLimit = '56kb',
    // urlencoded = true,
  },
  referrer: {
    // enabled: true,
    // directive: ['origin-when-cross-origin', 'strict-origin-when-cross-origin'],
  },
  requestId: {
    // enabled: true,
  },
  responseTime: env({
    default: {
      // enabled: true,
    },
    production: {
      enabled: false,
    },
  }),
  throttle: env({
    default: {
      // enabled: true,
      // duration: 900,
      // max: 450,
      // maxAuth: 450,
      // maxUnauth: 450,
    },
    test: {
      max: 10000,
    },
  }),
  version: {
    // enabled: true,
    availableVersions: [1],
  },
  xframe: {
    // enabled: true,
    // directive: 'deny',
  },
  xss: {
    // enabled: true,
    // mode: 'block',
  },
}));
