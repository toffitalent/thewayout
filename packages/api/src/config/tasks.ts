import { env, register } from '@two/config';

export const tasks = register('tasks', () => ({
  queue: {
    endpoint: env('AWS_ENDPOINT_URL'),
    region: env({
      default: 'us-west-2',
      env: 'AWS_REGION',
    }),
    url: env({
      default: '',
      test: '__QUEUE_URL__',
      env: 'API_TASKS_QUEUE_URL',
    }),
  },
}));
