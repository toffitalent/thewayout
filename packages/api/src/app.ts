import './config';

import { App, Config as AppConfig, Context } from '@disruptive-labs/server';
import config from '@two/config';
import { router } from '@app/routes';
import { onError } from '@app/services/Sentry';

const app = new App(config.get<AppConfig>('api.server'));

/* istanbul ignore next */
app.on('error', (error: Error, ctx?: Context) => {
  if (config.environment('staging', 'production')) {
    onError(error, ctx);
    return;
  }

  // eslint-disable-next-line no-console
  console.error(error);
});

/* istanbul ignore next */
app.on('stopped', () => {
  process.exit();
});

app.mount(router);

/* istanbul ignore if: avoid starting server if module imported for tests */
if (require.main === module) {
  process.on('SIGINT', () => app.stop());
  process.on('SIGTERM', () => app.stop());

  app.start();
}

export { app };
