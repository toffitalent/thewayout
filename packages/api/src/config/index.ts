import { load } from '@two/config';
import { auth } from './auth';
import { db } from './db';
import { server } from './server';
import { tasks } from './tasks';

export const apiConfig = load('api', [auth, db, server, tasks]);

export default apiConfig;
