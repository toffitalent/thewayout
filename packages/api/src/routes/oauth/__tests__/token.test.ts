import JWT from 'jsonwebtoken';
import config from '@two/config';
import { AuthGrantType } from '@two/shared';
import { app } from '@app/app';
import { App, AppClient, AppToken, AppUser } from '@app/models';
import {
  auth,
  createFakeApp,
  createFakeClient,
  fixtures,
  request,
  resetDb,
  resetStore,
} from '@test';

async function createApp() {
  const application = await App.query().insert(createFakeApp()).returning('*');
  const client = await AppClient.query()
    .insert(createFakeClient({ appId: application.id, grantType: AuthGrantType.AuthorizationCode }))
    .returning('*');

  await AppUser.query().insert({
    appId: application.id,
    userId: fixtures.user1.id,
    scope: ['user'],
  });

  return { application, client };
}

describe('Routes > OAuth > Token', () => {
  beforeEach(async () => {
    await resetDb();
    await resetStore();
  });

  describe('GET /api/v1/oauth/token', () => {
    test('returns token information', async () => {
      const res = await request().get('/api/v1/oauth/token').use(auth());

      expect(res.status).toBe(200);
      expect(res.type).toBe('application/json');
      expect(res.body).toMatchSnapshot({
        expiresAt: expect.any(Number),
        expiresIn: expect.any(Number),
      });
    });

    test('does not allow unauthenticated requests', async () => {
      const res = await request().get('/api/v1/oauth/token');
      expect(res.status).toBe(401);
      expect(res.type).toBe('application/json');
      expect(res.body).toMatchSnapshot();
    });
  });

  describe('POST /api/v1/oauth/token', () => {
    test('returns an invalid parameter response for unknown clients', async () => {
      const res = await request().post('/api/v1/oauth/token').send({
        clientId: fixtures.unknownId,
        clientSecret: '1234567890',
        grantType: 'password',
        username: fixtures.user1.email,
        password: 'DLabs2018!',
      });

      expect(res.status).toBe(400);
      expect(res.type).toBe('application/json');
      expect(res.body).toMatchSnapshot();
    });

    test('returns an invalid parameter response for invalid client secrets', async () => {
      const res = await request().post('/api/v1/oauth/token').send({
        clientId: fixtures.client.id,
        clientSecret: 'INVALID',
        grantType: 'password',
        username: fixtures.user1.email,
        password: 'DLabs2018!',
      });

      expect(res.status).toBe(400);
      expect(res.type).toBe('application/json');
      expect(res.body).toMatchSnapshot();
    });

    test('returns an invalid parameter response for mismatched grant type', async () => {
      const res = await request().post('/api/v1/oauth/token').send({
        clientId: fixtures.client.id,
        clientSecret: fixtures.client.secret,
        grantType: 'authorization_code',
        code: '12345',
      });

      expect(res.status).toBe(400);
      expect(res.type).toBe('application/json');
      expect(res.body).toMatchSnapshot();
    });

    describe('Authorization Code Grant', () => {
      let application: any;
      let client: any;

      beforeEach(async () => {
        ({ application, client } = await createApp());
      });

      test('returns access and refresh tokens', async () => {
        await app.store().set(`oauth:${client.id}:12345`, fixtures.user1.id);

        const res = await request().post('/api/v1/oauth/token').send({
          clientId: client.id,
          clientSecret: client.secret,
          grantType: 'authorization_code',
          code: '12345',
        });

        expect(res.status).toBe(200);
        expect(res.type).toBe('application/json');
        expect(res.body).toMatchSnapshot({
          accessToken: expect.any(String),
          refreshToken: expect.any(String),
        });

        const accessToken: any = JWT.verify(
          res.body.accessToken,
          config.get('api.auth.secret', ''),
        );

        expect(accessToken.azp).toBe(application.id);
        expect(accessToken.client_id).toBe(client.id);
        expect(accessToken.sub).toBe(fixtures.user1.id);
        expect(accessToken.scope).toBe('user');

        const refreshToken = await AppToken.query()
          .findOne({
            token: Buffer.from(res.body.refreshToken, 'base64'),
          })
          .throwIfNotFound();

        expect(refreshToken).not.toBeNull();
        expect(refreshToken.id).toBe(accessToken.jti);
        expect(refreshToken.appId).toBe(application.id);
        expect(refreshToken.clientId).toBe(client.id);
        expect(refreshToken.userId).toBe(fixtures.user1.id);
        expect(refreshToken.self).toBe(false);
      });

      test('deletes authorization code from cache to avoid replay', async () => {
        await app.store().set(`oauth:${client.id}:12345`, fixtures.user1.id);

        const res = await request().post('/api/v1/oauth/token').send({
          clientId: client.id,
          clientSecret: client.secret,
          grantType: 'authorization_code',
          code: '12345',
        });

        const keys = await app.store().keys(`oauth:${client.id}:*`);
        const value = await app.store().get(`oauth:${client.id}:12345`);

        expect(res.status).toBe(200);
        expect(keys).toHaveLength(0);
        expect(value).toBeNull();
      });

      test('returns an invalid parameter response for invalid codes', async () => {
        const res = await request().post('/api/v1/oauth/token').send({
          clientId: client.id,
          clientSecret: client.secret,
          grantType: 'authorization_code',
          code: '12345',
        });

        expect(res.status).toBe(400);
        expect(res.type).toBe('application/json');
        expect(res.body).toMatchSnapshot();
      });

      test('returns a not authorized response for unauthorized apps', async () => {
        await AppUser.query().where('appId', application.id).del();
        await app.store().set(`oauth:${client.id}:12345`, fixtures.user1.id);

        const res = await request().post('/api/v1/oauth/token').send({
          clientId: client.id,
          clientSecret: client.secret,
          grantType: 'authorization_code',
          code: '12345',
        });

        expect(res.status).toBe(403);
        expect(res.type).toBe('application/json');
        expect(res.body).toMatchSnapshot({ message: expect.any(String) });
        expect(res.body.message).toBe(`App ${application.id} is not authorized`);
      });
    });

    describe('Password Grant', () => {
      test('returns access and refresh tokens', async () => {
        const res = await request().post('/api/v1/oauth/token').send({
          clientId: fixtures.client.id,
          clientSecret: fixtures.client.secret,
          grantType: 'password',
          username: fixtures.user1.email,
          password: 'DLabs2018!',
        });

        expect(res.status).toBe(200);
        expect(res.type).toBe('application/json');
        expect(res.body).toMatchSnapshot({
          accessToken: expect.any(String),
          refreshToken: expect.any(String),
        });

        const accessToken: any = JWT.verify(
          res.body.accessToken,
          config.get('api.auth.secret', ''),
        );

        expect(accessToken.azp).toBe(fixtures.app.id);
        expect(accessToken.client_id).toBe(fixtures.client.id);
        expect(accessToken.sub).toBe(fixtures.user1.id);
        expect(accessToken.scope).toBe('self client');

        const refreshToken = await AppToken.query()
          .findOne({
            token: Buffer.from(res.body.refreshToken, 'base64'),
          })
          .throwIfNotFound();

        expect(refreshToken).not.toBeNull();
        expect(refreshToken.id).toBe(accessToken.jti);
        expect(refreshToken.appId).toBe(fixtures.app.id);
        expect(refreshToken.clientId).toBe(fixtures.client.id);
        expect(refreshToken.userId).toBe(fixtures.user1.id);
        expect(refreshToken.self).toBe(true);
      });

      test('returns an invalid credentials response for invalid usernames/passwords', async () => {
        const res = await request().post('/api/v1/oauth/token').send({
          clientId: fixtures.client.id,
          clientSecret: fixtures.client.secret,
          grantType: 'password',
          username: fixtures.user1.email,
          password: 'invalid_password',
        });

        expect(res.status).toBe(401);
        expect(res.type).toBe('application/json');
        expect(res.body).toMatchSnapshot();
      });
    });

    describe('Refresh Token Grant', () => {
      test('returns access and refresh tokens', async () => {
        const refresh = await AppToken.query()
          .insert({
            appId: fixtures.app.id,
            clientId: fixtures.client.id,
            userId: fixtures.user1.id,
            self: true,
          })
          .returning('*');

        const res = await request()
          .post('/api/v1/oauth/token')
          .send({
            clientId: fixtures.client.id,
            clientSecret: fixtures.client.secret,
            grantType: 'refresh_token',
            refreshToken: refresh.token.toString('base64'),
          });

        expect(res.status).toBe(200);
        expect(res.type).toBe('application/json');
        expect(res.body).toMatchSnapshot({
          accessToken: expect.any(String),
          refreshToken: expect.any(String),
        });

        const accessToken: any = JWT.verify(
          res.body.accessToken,
          config.get('api.auth.secret', ''),
        );

        expect(accessToken.azp).toBe(fixtures.app.id);
        expect(accessToken.client_id).toBe(fixtures.client.id);
        expect(accessToken.sub).toBe(fixtures.user1.id);
        expect(accessToken.scope).toBe('self client');

        const refreshToken = await AppToken.query()
          .findOne({
            token: Buffer.from(res.body.refreshToken, 'base64'),
          })
          .throwIfNotFound();

        expect(refreshToken).not.toBeNull();
        expect(refreshToken.id).toBe(accessToken.jti);
        expect(refreshToken.appId).toBe(fixtures.app.id);
        expect(refreshToken.clientId).toBe(fixtures.client.id);
        expect(refreshToken.userId).toBe(fixtures.user1.id);
        expect(refreshToken.self).toBe(true);
      });

      test('only includes authorized scopes for non-password-grant tokens', async () => {
        const { application, client } = await createApp();
        const refresh = await AppToken.query()
          .insert({
            appId: application.id,
            clientId: client.id,
            userId: fixtures.user1.id,
            self: false,
          })
          .returning('*');

        const res = await request()
          .post('/api/v1/oauth/token')
          .send({
            clientId: client.id,
            clientSecret: client.secret,
            grantType: 'refresh_token',
            refreshToken: refresh.token.toString('base64'),
          });

        expect(res.status).toBe(200);
        expect(res.type).toBe('application/json');
        expect(res.body).toMatchSnapshot({
          accessToken: expect.any(String),
          refreshToken: expect.any(String),
        });

        const accessToken: any = JWT.verify(
          res.body.accessToken,
          config.get('api.auth.secret', ''),
        );

        expect(accessToken.azp).toBe(application.id);
        expect(accessToken.client_id).toBe(client.id);
        expect(accessToken.sub).toBe(fixtures.user1.id);
        expect(accessToken.scope).toBe('user');

        const refreshToken = await AppToken.query()
          .findOne({
            token: Buffer.from(res.body.refreshToken, 'base64'),
          })
          .throwIfNotFound();

        expect(refreshToken).not.toBeNull();
        expect(refreshToken.id).toBe(accessToken.jti);
        expect(refreshToken.appId).toBe(application.id);
        expect(refreshToken.clientId).toBe(client.id);
        expect(refreshToken.userId).toBe(fixtures.user1.id);
        expect(refreshToken.self).toBe(false);
      });

      test('returns an invalid parameter response for invalid tokens', async () => {
        const res = await request()
          .post('/api/v1/oauth/token')
          .send({
            clientId: fixtures.client.id,
            clientSecret: fixtures.client.secret,
            grantType: 'refresh_token',
            refreshToken: Buffer.from('000000000').toString('base64'),
          });

        expect(res.status).toBe(400);
        expect(res.type).toBe('application/json');
        expect(res.body).toMatchSnapshot();
      });
    });
  });

  describe('DELETE /api/v1/oauth/token', () => {
    test('removes the associated refresh token', async () => {
      const refresh = await AppToken.query()
        .insert({
          appId: fixtures.app.id,
          clientId: fixtures.client.id,
          userId: fixtures.user1.id,
        })
        .returning('*');

      const payload = {
        jti: refresh.id,
        sub: fixtures.user1.id,
        azp: fixtures.app.id,
        client_id: fixtures.client.id,
        scope: '',
      };

      const accessToken = JWT.sign(payload, config.get<string>('api.auth.secret', ''), {
        algorithm: config.get('api.auth.algorithm', 'HS256'),
        audience: config.get('api.auth.audience', 'api'),
        issuer: config.get('api.auth.issuer', 'api'),
        expiresIn: 10,
      });

      const res = await request()
        .delete('/api/v1/oauth/token')
        .set('Authorization', `Bearer ${accessToken}`);

      const refreshToken = await AppToken.query().findById(refresh.id);

      expect(res.status).toBe(204);
      expect(refreshToken).toBeUndefined();
    });
  });
});
