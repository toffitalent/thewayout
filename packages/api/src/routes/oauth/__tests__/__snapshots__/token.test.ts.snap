// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes > OAuth > Token GET /api/v1/oauth/token does not allow unauthenticated requests 1`] = `
{
  "code": "InvalidCredentials",
  "message": "Authentication required",
  "status": 401,
}
`;

exports[`Routes > OAuth > Token GET /api/v1/oauth/token returns token information 1`] = `
{
  "appId": "01817e24-0d42-79ee-bfde-bc16bf2b0e77",
  "clientId": "01817e24-0d42-7adb-aa85-cd0c708ca3be",
  "expiresAt": Any<Number>,
  "expiresIn": Any<Number>,
  "roles": [
    "self",
  ],
  "scope": "self user",
  "userId": "01817e24-0d42-7176-b613-889fe94905f0",
}
`;

exports[`Routes > OAuth > Token POST /api/v1/oauth/token Authorization Code Grant returns a not authorized response for unauthorized apps 1`] = `
{
  "code": "NotAuthorized",
  "message": Any<String>,
  "status": 403,
}
`;

exports[`Routes > OAuth > Token POST /api/v1/oauth/token Authorization Code Grant returns access and refresh tokens 1`] = `
{
  "accessToken": Any<String>,
  "refreshToken": Any<String>,
  "scope": "user",
  "tokenType": "Bearer",
}
`;

exports[`Routes > OAuth > Token POST /api/v1/oauth/token Authorization Code Grant returns an invalid parameter response for invalid codes 1`] = `
{
  "code": "InvalidRequest",
  "errors": [
    {
      "error": "invalid",
      "key": [
        "code",
      ],
      "source": "body",
    },
  ],
  "message": "Authorization code is invalid",
  "status": 400,
}
`;

exports[`Routes > OAuth > Token POST /api/v1/oauth/token Password Grant returns access and refresh tokens 1`] = `
{
  "accessToken": Any<String>,
  "refreshToken": Any<String>,
  "roles": [
    "client",
  ],
  "scope": "self client",
  "tokenType": "Bearer",
}
`;

exports[`Routes > OAuth > Token POST /api/v1/oauth/token Password Grant returns an invalid credentials response for invalid usernames/passwords 1`] = `
{
  "code": "InvalidCredentials",
  "message": "Invalid username/password",
  "status": 401,
}
`;

exports[`Routes > OAuth > Token POST /api/v1/oauth/token Refresh Token Grant only includes authorized scopes for non-password-grant tokens 1`] = `
{
  "accessToken": Any<String>,
  "refreshToken": Any<String>,
  "scope": "user",
  "tokenType": "Bearer",
}
`;

exports[`Routes > OAuth > Token POST /api/v1/oauth/token Refresh Token Grant returns access and refresh tokens 1`] = `
{
  "accessToken": Any<String>,
  "refreshToken": Any<String>,
  "roles": [
    "client",
  ],
  "scope": "self client",
  "tokenType": "Bearer",
}
`;

exports[`Routes > OAuth > Token POST /api/v1/oauth/token Refresh Token Grant returns an invalid parameter response for invalid tokens 1`] = `
{
  "code": "InvalidRequest",
  "errors": [
    {
      "error": "invalid",
      "key": [
        "refreshToken",
      ],
      "source": "body",
    },
  ],
  "message": "Refresh token is invalid",
  "status": 400,
}
`;

exports[`Routes > OAuth > Token POST /api/v1/oauth/token returns an invalid parameter response for invalid client secrets 1`] = `
{
  "code": "InvalidRequest",
  "errors": [
    {
      "error": "invalid",
      "key": [
        "clientSecret",
      ],
      "source": "body",
    },
  ],
  "message": "Client secret does not match",
  "status": 400,
}
`;

exports[`Routes > OAuth > Token POST /api/v1/oauth/token returns an invalid parameter response for mismatched grant type 1`] = `
{
  "code": "InvalidRequest",
  "errors": [
    {
      "error": "invalid",
      "key": [
        "grantType",
      ],
      "source": "body",
    },
  ],
  "message": "Client not authorized for authorization_code grant",
  "status": 400,
}
`;

exports[`Routes > OAuth > Token POST /api/v1/oauth/token returns an invalid parameter response for unknown clients 1`] = `
{
  "code": "InvalidRequest",
  "errors": [
    {
      "error": "invalid",
      "key": [
        "clientId",
      ],
      "source": "body",
    },
  ],
  "message": "Client 00000000-0000-0000-0000-000000000000 not found",
  "status": 400,
}
`;
